File: apps/edge-worker/src/services/ai-key-crypto.ts
const ENCRYPTION_VERSION = 'aes-gcm-v1'
const AES_GCM_IV_BYTES = 12

export type AiApiKeyEncryptedPayload = {
  v: typeof ENCRYPTION_VERSION
  iv: string
  data: string
}

type StoredAiApiKeyFields = {
  ai_api_key?: string | null
  ai_api_key_encrypted?: string | null
}

function normalizeSecret(secret: string | null | undefined): string {
  return String(secret || '').trim()
}

function bytesToBase64(bytes: Uint8Array): string {
  let binary = ''
  bytes.forEach((byte) => {
    binary += String.fromCharCode(byte)
  })
  return btoa(binary)
}

function base64ToBytes(value: string): Uint8Array {
  const binary = atob(value)
  const bytes = new Uint8Array(binary.length)
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index)
  }
  return bytes
}

async function deriveAesKey(secret: string): Promise<CryptoKey> {
  const normalized = normalizeSecret(secret)
  if (!normalized) {
    throw new Error('AI_KEY_ENCRYPTION_SECRET 未配置，无法安全保存 API Key。')
  }

  const material = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(normalized))
  return crypto.subtle.importKey('raw', material, 'AES-GCM', false, ['encrypt', 'decrypt'])
}

export async function encryptAiApiKey(apiKey: string, secret: string | null | undefined): Promise<string> {
  const normalizedKey = String(apiKey || '').trim()
  if (!normalizedKey) {
    throw new Error('API Key 不能为空。')
  }

  const key = await deriveAesKey(secret || '')
  const iv = crypto.getRandomValues(new Uint8Array(AES_GCM_IV_BYTES))
  const encrypted = await crypto.subtle.encrypt(
    {
      name: 'AES-GCM',
      iv,
    },
    key,
    new TextEncoder().encode(normalizedKey)
  )

  const payload: AiApiKeyEncryptedPayload = {
    v: ENCRYPTION_VERSION,
    iv: bytesToBase64(iv),
    data: bytesToBase64(new Uint8Array(encrypted)),
  }

  return JSON.stringify(payload)
}

export async function decryptAiApiKey(
  encryptedValue: string,
  secret: string | null | undefined
): Promise<string | null> {
  const normalizedEncryptedValue = String(encryptedValue || '').trim()
  if (!normalizedEncryptedValue) return null

  const key = await deriveAesKey(secret || '')
  const payload = JSON.parse(normalizedEncryptedValue) as Partial<AiApiKeyEncryptedPayload>
  if (payload.v !== ENCRYPTION_VERSION || !payload.iv || !payload.data) {
    throw new Error('AI API Key 加密载荷版本不受支持。')
  }

  const decrypted = await crypto.subtle.decrypt(
    {
      name: 'AES-GCM',
      iv: base64ToBytes(payload.iv),
    },
    key,
    base64ToBytes(payload.data)
  )

  return new TextDecoder().decode(decrypted)
}

export async function resolveStoredAiApiKey(
  settings: StoredAiApiKeyFields | null | undefined,
  secret: string | null | undefined
): Promise<string | null> {
  const encryptedValue = String(settings?.ai_api_key_encrypted || '').trim()
  if (encryptedValue) {
    try {
      return await decryptAiApiKey(encryptedValue, secret)
    } catch (error) {
      console.warn('Decrypt user AI API key failed; ignoring encrypted key:', error)
      return null
    }
  }

  const legacyPlaintextKey = String(settings?.ai_api_key || '').trim()
  return legacyPlaintextKey || null
}

export const AI_API_KEY_ENCRYPTION_VERSION = ENCRYPTION_VERSION

File: app/services/ai_key_crypto.py
from __future__ import annotations

import base64
import hashlib
import json
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from app.config import settings


AI_API_KEY_ENCRYPTION_VERSION = "aes-gcm-v1"


def _normalize_secret(secret: str | None) -> str:
    return str(secret or "").strip()


def decrypt_ai_api_key(encrypted_value: str, secret: str | None = None) -> str | None:
    normalized_value = str(encrypted_value or "").strip()
    if not normalized_value:
        return None

    normalized_secret = _normalize_secret(secret if secret is not None else settings.AI_KEY_ENCRYPTION_SECRET)
    if not normalized_secret:
        return None

    payload: dict[str, Any] = json.loads(normalized_value)
    if payload.get("v") != AI_API_KEY_ENCRYPTION_VERSION:
        raise ValueError("AI API Key 加密载荷版本不受支持。")

    iv = base64.b64decode(str(payload.get("iv") or ""))
    data = base64.b64decode(str(payload.get("data") or ""))
    key = hashlib.sha256(normalized_secret.encode("utf-8")).digest()
    decrypted = AESGCM(key).decrypt(iv, data, None)
    return decrypted.decode("utf-8")


def resolve_stored_ai_api_key(row: dict[str, Any], secret: str | None = None) -> str:
    encrypted_value = str(row.get("ai_api_key_encrypted") or "").strip()
    if encrypted_value:
        decrypted = decrypt_ai_api_key(encrypted_value, secret)
        if decrypted:
            return decrypted

    return str(row.get("ai_api_key") or "").strip()
